pt_miniscreen/app.py

```diff
@@ -88,59 +98,63 @@ def setup_events(self) -> None:
+        self.on_button_released()
 
-        class MiniscreenButton(Enum):
-            UP = auto()
-            DOWN = auto()
-            SELECT = auto()
-            CANCEL = auto()
+        current_tile = self.current_tile_group.current_menu_tile
+        if isinstance(current_tile, PaginatedTile):
+            current_tile.scroll_to_next()
 
-        def handle_button_press(button: MiniscreenButton):
-            logger.debug(
-                f"Handling button {button} for tile group {self.current_tile_group}"
-            )
+    def on_select_button_released(self):
+        logger.debug(f"Handling select button for tile group {self.current_tile_group}")
+        self.on_button_released()
 
-            if not self.state_manager.buttons_should_be_handled():
-                logger.info("State manager says that buttons should not be handled")
+        current_tile = self.current_tile_group.current_menu_tile
+        if isinstance(current_tile, MenuTile):
+            handled = current_tile.go_to_child_menu()
+            if handled:
                 return
 
-            self.state_manager.wake()
+        # if menu doesn't have a child menu it might be an action page
+        current_tile.start_action()
 
-            handler = {
-                MiniscreenButton.CANCEL: self.current_tile_group.handle_cancel_btn,
-                MiniscreenButton.SELECT: self.current_tile_group.handle_select_btn,
-                MiniscreenButton.UP: self.current_tile_group.handle_up_btn,
-                MiniscreenButton.DOWN: self.current_tile_group.handle_down_btn,
-            }[button]
+    def on_cancel_button_released(self):
+        logger.debug(f"Handling cancel button for tile group {self.current_tile_group}")
+        self.on_button_released()
 
-            if not handler() and button == MiniscreenButton.CANCEL:
-                logger.debug(
-                    "Button press not handled by current tile group - going to next tile group"
-                )
-                self.pop_tile_group()
+        if hasattr(self.current_tile_group, "handle_cancel_btn") and callable(
+            self.current_tile_group.handle_cancel_btn
+        ):
+            handled = self.current_tile_group.handle_cancel_btn()
+            if handled:
+                return
 
-        self.miniscreen.cancel_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.CANCEL
-        )
-        self.miniscreen.select_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.SELECT
-        )
-        self.miniscreen.up_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.UP
-        )
-        self.miniscreen.down_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.DOWN
-        )
+        logger.debug("Button press not handled by current tile group - popping group")
+        self.pop_tile_group()
```

Comment:
I don't really understand why the original code needed this extra case of doing `pop_tile_group` if the tile won't handle the action...
I think the current tiley menuy groupy thingy should handle popping it's own sub menu with `current_tile_group.handle_cancel_btn` right
and at this app level, the functionality to pop when cancel is pressed isn't required?

IMO we the previous code here was simpler, deferring to the current tile button handlers, and it just needed to be simplified further by removing this extra case...

Reply:
Popping a TileGroup is higher level than the can be handled in the TileGroup itself. I think you are confusing the children of the menus with the TileGroups themselves, which are things like the screensaver and the bootsplash.

Reply:
Right, this is popping the app level tile group, eg screensaver -> hudmenu

I wasn't saying that should be handled by the current tile group (I was saying that almost everything else could be). I was saying this isn't required at all... When does pressing cancel need to trigger a pop of the top level tile groups?

Reply:
It doesn't pop the top level one, but if two TileGroups are in the stack, and neither of them handle the button press, then the app pops the top one on the stack, the screensaver for example.

This would be much easier to reason about if the app explicitly handled the cancel button ;)